Const correctness in IntermediateIndex

Consider:

```c++
#include "../UMESimd.h"

using namespace UME::SIMD;

template <typename T1, typename T2>
bool compare(T1 const & value1, T2 const & value2)
{
    if (value1 == value2) {
        std::cout << "true" << std::endl;
        return true;
    } else {
        std::cout << "false" << std::endl;
        return false;
    }
}

int main()
{
    SIMD4_64u a(5, 3, 8, 4), b(13, 984, 5, 0);

    // won't work
    compare(a[0], b[2]);

    // works
    if (a[0] == b[2]) {
        std::cout << "true" << std::endl;
    } else {
        std::cout << "false" << std::endl;
    }
    return 0;
}

```

This will produce 

gcc:
```terminal
index_bug.cpp: In instantiation of ‘bool compare(const T1&, const T2&) [with T1 = UME::SIMD::IntermediateIndex<UME::SIMD::SIMDVec_u<long unsigned int, 4u>, long unsigned int>; T2 = UME::SIMD::IntermediateIndex<UME::SIMD::SIMDVec_u<long unsigned int, 4u>, long unsigned int>]’:
index_bug.cpp:22:23:   required from here
index_bug.cpp:8:16: error: no match for ‘operator==’ (operand types are ‘const UME::SIMD::IntermediateIndex<UME::SIMD::SIMDVec_u<long unsigned int, 4u>, long unsigned int>’ and ‘const UME::SIMD::IntermediateIndex<UME::SIMD::SIMDVec_u<long unsigned int, 4u>, long unsigned int>’)
     if (value1 == value2) {
         ~~~~~~~^~~~~~~~~
index_bug.cpp:8:16: note: candidate: operator==(long unsigned int, long unsigned int) <built-in>
index_bug.cpp:8:16: note:   conversion of argument 2 would be ill-formed:
index_bug.cpp:8:16: error: passing ‘const UME::SIMD::IntermediateIndex<UME::SIMD::SIMDVec_u<long unsigned int, 4u>, long unsigned int>’ as ‘this’ argument discards qualifiers [-fpermissive]
In file included from ../plugins/UMESimdPluginScalarEmulation.h:36:0,
                 from ../UMESimd.h:147,
                 from index_bug.cpp:1:
../plugins/../UMESimdInterface.h:654:26: note:   in call to ‘UME::SIMD::IntermediateIndex<VEC_TYPE, SCALAR_TYPE>::operator SCALAR_TYPE() [with VEC_TYPE = UME::SIMD::SIMDVec_u<long unsigned int, 4u>; SCALAR_TYPE = long unsigned int]’
         UME_FORCE_INLINE operator SCALAR_TYPE() { return mVecRef_RW.extract(mIndexRef); }
                          ^~~~~~~~

[...]
```

clang:
```terminal
index_bug.cpp:8:16: error: invalid operands to binary expression ('const UME::SIMD::IntermediateIndex<UME::SIMD::SIMDVec_u<unsigned long, 4>, unsigned
      long>' and 'const UME::SIMD::IntermediateIndex<UME::SIMD::SIMDVec_u<unsigned long, 4>, unsigned long>')
    if (value1 == value2) {
        ~~~~~~ ^  ~~~~~~
index_bug.cpp:22:5: note: in instantiation of function template specialization 'compare<UME::SIMD::IntermediateIndex<UME::SIMD::SIMDVec_u<unsigned
      long, 4>, unsigned long>, UME::SIMD::IntermediateIndex<UME::SIMD::SIMDVec_u<unsigned long, 4>, unsigned long> >' requested here
    compare(a[0], b[2]);
    ^
1 error generated.
```

----

The problem is that all comparison operators are not declared as const. Furthermore, the templated function won't allow implicit type conversion of the second argument.

--- a/UMESimdInterface.h
+++ b/UMESimdInterface.h
@@ -665,15 +665,15 @@ namespace SIMD
         }
         UME_FORCE_INLINE bool operator== (IntermediateIndex const & x) {
             return mVecRef_RW.extract(mIndexRef) ==
-                x.mVecRef_RW.extract(mIndexRef);
+                x.mVecRef_RW.extract(x.mIndexRef);
         }
         template<typename T>
         UME_FORCE_INLINE bool operator!=(T const & rhs) {
-            return mVecRef_RW.extract(mIndexRef) != rhs;
+            return mVecRef_RW.extract(mIndexRef) != SCALAR_TYPE(rhs);
         }
         UME_FORCE_INLINE bool operator!= (IntermediateIndex const & x) {
             return mVecRef_RW.extract(mIndexRef) !=
-                x.mVecRef_RW.extract(mIndexRef);
+                x.mVecRef_RW.extract(x.mIndexRef);
         }
         template<typename T>
         UME_FORCE_INLINE SCALAR_TYPE operator+ (T const & x) {
@@ -681,7 +681,7 @@ namespace SIMD
         }
         UME_FORCE_INLINE SCALAR_TYPE operator+ (IntermediateIndex const & x) {
             return mVecRef_RW.extract(mIndexRef) +
-                x.mVecRef_RW.extract(mIndexRef);
+                x.mVecRef_RW.extract(x.mIndexRef);
         }
         template<typename T>
         UME_FORCE_INLINE SCALAR_TYPE operator- (T const & x) {
@@ -689,7 +689,7 @@ namespace SIMD
         }
         UME_FORCE_INLINE SCALAR_TYPE operator- (IntermediateIndex const & x) {
             return mVecRef_RW.extract(mIndexRef) -
-                x.mVecRef_RW.extract(mIndexRef);
+                x.mVecRef_RW.extract(x.mIndexRef);
         }
         template<typename T>
         UME_FORCE_INLINE SCALAR_TYPE operator* (T const & x) {
@@ -697,7 +697,7 @@ namespace SIMD
         }
         UME_FORCE_INLINE SCALAR_TYPE operator* (IntermediateIndex const & x) {
             return mVecRef_RW.extract(mIndexRef) *
-                x.mVecRef_RW.extract(mIndexRef);
+                x.mVecRef_RW.extract(x.mIndexRef);
         }
         template<typename T>
         UME_FORCE_INLINE SCALAR_TYPE operator/ (T const & x) {
@@ -705,7 +705,7 @@ namespace SIMD
         }
         UME_FORCE_INLINE SCALAR_TYPE operator/ (IntermediateIndex const & x) {
             return mVecRef_RW.extract(mIndexRef) /
-                x.mVecRef_RW.extract(mIndexRef);
+                x.mVecRef_RW.extract(x.mIndexRef);
         }
         template<typename T>
         UME_FORCE_INLINE SCALAR_TYPE operator% (T const & x) {
@@ -713,7 +713,7 @@ namespace SIMD
         }
         UME_FORCE_INLINE SCALAR_TYPE operator% (IntermediateIndex const & x) {
             return mVecRef_RW.extract(mIndexRef) %
-                x.mVecRef_RW.extract(mIndexRef);
+                x.mVecRef_RW.extract(x.mIndexRef);
         }
         template<typename T>
         UME_FORCE_INLINE SCALAR_TYPE operator& (T const & x) {
@@ -721,7 +721,7 @@ namespace SIMD
         }
         UME_FORCE_INLINE SCALAR_TYPE operator& (IntermediateIndex const & x) {
             return mVecRef_RW.extract(mIndexRef) &
-                x.mVecRef_RW.extract(mIndexRef);
+                x.mVecRef_RW.extract(x.mIndexRef);
         }
         template<typename T>
         UME_FORCE_INLINE SCALAR_TYPE operator| (T const & x) {
@@ -729,7 +729,7 @@ namespace SIMD
         }
         UME_FORCE_INLINE SCALAR_TYPE operator| (IntermediateIndex const & x) {
             return mVecRef_RW.extract(mIndexRef) |
-                x.mVecRef_RW.extract(mIndexRef);
+                x.mVecRef_RW.extract(x.mIndexRef);
         }
         template<typename T>
         UME_FORCE_INLINE SCALAR_TYPE operator^ (T const & x) {
@@ -737,7 +737,7 @@ namespace SIMD
         }
         UME_FORCE_INLINE SCALAR_TYPE operator^ (IntermediateIndex const & x) {
             return mVecRef_RW.extract(mIndexRef) ^
-                x.mVecRef_RW.extract(mIndexRef);
+                x.mVecRef_RW.extract(x.mIndexRef);
         }
         template<typename T>
         UME_FORCE_INLINE SCALAR_TYPE operator<< (T const & x) {
@@ -745,7 +745,7 @@ namespace SIMD
         }
         UME_FORCE_INLINE SCALAR_TYPE operator<< (IntermediateIndex const & x) {
             return mVecRef_RW.extract(mIndexRef) <<
-                x.mVecRef_RW.extract(mIndexRef);
+                x.mVecRef_RW.extract(x.mIndexRef);
         }
         template<typename T>
         UME_FORCE_INLINE SCALAR_TYPE operator>> (T const & x) {
@@ -753,7 +753,7 @@ namespace SIMD
         }
         UME_FORCE_INLINE SCALAR_TYPE operator>> (IntermediateIndex const & x) {
             return mVecRef_RW.extract(mIndexRef) >>
-                x.mVecRef_RW.extract(mIndexRef);
+                x.mVecRef_RW.extract(x.mIndexRef);
         }
 
     private:

--- a/UMESimdInterface.h
+++ b/UMESimdInterface.h
@@ -651,7 +651,7 @@ namespace SIMD
             mVecRef_RW.insert(mIndexRef, mVecRef_RW[mIndexRef] >> scalarRhs);
         }
 
-        UME_FORCE_INLINE operator SCALAR_TYPE() { return mVecRef_RW.extract(mIndexRef); }
+        UME_FORCE_INLINE operator SCALAR_TYPE() const { return mVecRef_RW.extract(mIndexRef); }
 
         // Comparison operators accept any type of scalar to allow mixing 
         // scalar types.
@@ -660,98 +660,98 @@ namespace SIMD
             typename = typename std::enable_if<std::is_fundamental<T>::value, void*>::type
             >
         UME_FORCE_INLINE bool operator==(
-                T const & rhs) {
+                T const & rhs) const {
             return mVecRef_RW.extract(mIndexRef) == SCALAR_TYPE(rhs);
         }
-        UME_FORCE_INLINE bool operator== (IntermediateIndex const & x) {
+        UME_FORCE_INLINE bool operator== (IntermediateIndex const & x) const {
             return mVecRef_RW.extract(mIndexRef) ==
                 x.mVecRef_RW.extract(x.mIndexRef);
         }
         template<typename T>
-        UME_FORCE_INLINE bool operator!=(T const & rhs) {
+        UME_FORCE_INLINE bool operator!=(T const & rhs) const {
             return mVecRef_RW.extract(mIndexRef) != SCALAR_TYPE(rhs);
         }
-        UME_FORCE_INLINE bool operator!= (IntermediateIndex const & x) {
+        UME_FORCE_INLINE bool operator!= (IntermediateIndex const & x) const {
             return mVecRef_RW.extract(mIndexRef) !=
                 x.mVecRef_RW.extract(x.mIndexRef);
         }
         template<typename T>
-        UME_FORCE_INLINE SCALAR_TYPE operator+ (T const & x) {
+        UME_FORCE_INLINE SCALAR_TYPE operator+ (T const & x) const {
             return mVecRef_RW.extract(mIndexRef) + SCALAR_TYPE(x);
         }
-        UME_FORCE_INLINE SCALAR_TYPE operator+ (IntermediateIndex const & x) {
+        UME_FORCE_INLINE SCALAR_TYPE operator+ (IntermediateIndex const & x) const {
             return mVecRef_RW.extract(mIndexRef) +
                 x.mVecRef_RW.extract(x.mIndexRef);
         }
         template<typename T>
-        UME_FORCE_INLINE SCALAR_TYPE operator- (T const & x) {
+        UME_FORCE_INLINE SCALAR_TYPE operator- (T const & x) const {
             return mVecRef_RW.extract(mIndexRef) - SCALAR_TYPE(x);
         }
-        UME_FORCE_INLINE SCALAR_TYPE operator- (IntermediateIndex const & x) {
+        UME_FORCE_INLINE SCALAR_TYPE operator- (IntermediateIndex const & x) const {
             return mVecRef_RW.extract(mIndexRef) -
                 x.mVecRef_RW.extract(x.mIndexRef);
         }
         template<typename T>
-        UME_FORCE_INLINE SCALAR_TYPE operator* (T const & x) {
+        UME_FORCE_INLINE SCALAR_TYPE operator* (T const & x) const {
             return mVecRef_RW.extract(mIndexRef) * SCALAR_TYPE(x);
         }
-        UME_FORCE_INLINE SCALAR_TYPE operator* (IntermediateIndex const & x) {
+        UME_FORCE_INLINE SCALAR_TYPE operator* (IntermediateIndex const & x) const {
             return mVecRef_RW.extract(mIndexRef) *
                 x.mVecRef_RW.extract(x.mIndexRef);
         }
         template<typename T>
-        UME_FORCE_INLINE SCALAR_TYPE operator/ (T const & x) {
+        UME_FORCE_INLINE SCALAR_TYPE operator/ (T const & x) const {
             return mVecRef_RW.extract(mIndexRef) / SCALAR_TYPE(x);
         }
-        UME_FORCE_INLINE SCALAR_TYPE operator/ (IntermediateIndex const & x) {
+        UME_FORCE_INLINE SCALAR_TYPE operator/ (IntermediateIndex const & x) const {
             return mVecRef_RW.extract(mIndexRef) /
                 x.mVecRef_RW.extract(x.mIndexRef);
         }
         template<typename T>
-        UME_FORCE_INLINE SCALAR_TYPE operator% (T const & x) {
+        UME_FORCE_INLINE SCALAR_TYPE operator% (T const & x) const {
             return mVecRef_RW.extract(mIndexRef) % SCALAR_TYPE(x);
         }
-        UME_FORCE_INLINE SCALAR_TYPE operator% (IntermediateIndex const & x) {
+        UME_FORCE_INLINE SCALAR_TYPE operator% (IntermediateIndex const & x) const {
             return mVecRef_RW.extract(mIndexRef) %
                 x.mVecRef_RW.extract(x.mIndexRef);
         }
         template<typename T>
-        UME_FORCE_INLINE SCALAR_TYPE operator& (T const & x) {
+        UME_FORCE_INLINE SCALAR_TYPE operator& (T const & x) const {
             return mVecRef_RW.extract(mIndexRef) & SCALAR_TYPE(x);
         }
-        UME_FORCE_INLINE SCALAR_TYPE operator& (IntermediateIndex const & x) {
+        UME_FORCE_INLINE SCALAR_TYPE operator& (IntermediateIndex const & x) const {
             return mVecRef_RW.extract(mIndexRef) &
                 x.mVecRef_RW.extract(x.mIndexRef);
         }
         template<typename T>
-        UME_FORCE_INLINE SCALAR_TYPE operator| (T const & x) {
+        UME_FORCE_INLINE SCALAR_TYPE operator| (T const & x) const {
             return mVecRef_RW.extract(mIndexRef) | SCALAR_TYPE(x);
         }
-        UME_FORCE_INLINE SCALAR_TYPE operator| (IntermediateIndex const & x) {
+        UME_FORCE_INLINE SCALAR_TYPE operator| (IntermediateIndex const & x) const {
             return mVecRef_RW.extract(mIndexRef) |
                 x.mVecRef_RW.extract(x.mIndexRef);
         }
         template<typename T>
-        UME_FORCE_INLINE SCALAR_TYPE operator^ (T const & x) {
+        UME_FORCE_INLINE SCALAR_TYPE operator^ (T const & x) const {
             return mVecRef_RW.extract(mIndexRef) ^ SCALAR_TYPE(x);
         }
-        UME_FORCE_INLINE SCALAR_TYPE operator^ (IntermediateIndex const & x) {
+        UME_FORCE_INLINE SCALAR_TYPE operator^ (IntermediateIndex const & x) const {
             return mVecRef_RW.extract(mIndexRef) ^
                 x.mVecRef_RW.extract(x.mIndexRef);
         }
         template<typename T>
-        UME_FORCE_INLINE SCALAR_TYPE operator<< (T const & x) {
+        UME_FORCE_INLINE SCALAR_TYPE operator<< (T const & x) const {
             return mVecRef_RW.extract(mIndexRef) << SCALAR_TYPE(x);
         }
-        UME_FORCE_INLINE SCALAR_TYPE operator<< (IntermediateIndex const & x) {
+        UME_FORCE_INLINE SCALAR_TYPE operator<< (IntermediateIndex const & x) const {
             return mVecRef_RW.extract(mIndexRef) <<
                 x.mVecRef_RW.extract(x.mIndexRef);
         }
         template<typename T>
-        UME_FORCE_INLINE SCALAR_TYPE operator>> (T const & x) {
+        UME_FORCE_INLINE SCALAR_TYPE operator>> (T const & x) const {
             return mVecRef_RW.extract(mIndexRef) >> SCALAR_TYPE(x);
         }
-        UME_FORCE_INLINE SCALAR_TYPE operator>> (IntermediateIndex const & x) {
+        UME_FORCE_INLINE SCALAR_TYPE operator>> (IntermediateIndex const & x) const {
             return mVecRef_RW.extract(mIndexRef) >>
                 x.mVecRef_RW.extract(x.mIndexRef);
         }

